### PR TITLE
Fix reads checks for complex assignment and increment/decrement for immutable variables.

### LINKED ISSUE #9196: [codegen] ICE because "memory offset of unknown immutable queried"
## Description

```
contract A {
  int immutable a;
  constructor() public { a = 5; }
  function f() public { (--a); }
}
```

throws

https://github.com/ethereum/solidity/blob/21a9d3dd21ec0bb31fffb0daa0de902e274d7850/libsolidity/codegen/CompilerContext.cpp#L86

## Environment

- Compiler version: latest develop

## Steps to Reproduce

```
$ solc <input.sol>
```

### LINKED ISSUE #9550: ICE in solidity::frontend::ImmutableItem::setToZero(const solidity::langutil::SourceLocation &, bool) const
## Description

This contract:
```solidity
contract C
{
    uint public immutable a = 4;
    function f() public {
	    delete a;
    }
}
```

produces:

```
Internal compiler error during compilation:
/home/user/solidity/libsolidity/codegen/LValue.cpp(181): Throw in function virtual void solidity::frontend::ImmutableItem::setToZero(const solidity::langutil::SourceLocation &, bool) const
Dynamic exception type: boost::exception_detail::clone_impl<solidity::langutil::InternalCompilerError>
std::exception::what: Attempted to set immutable variable to zero.
[solidity::util::tag_comment*] = Attempted to set immutable variable to zero.
```

when compiled with solc

On master, using AFL fuzzing.  Another discovery using https://github.com/agroce/afl-compiler-fuzzer.

## Environment

- Compiler version:  0.7.1-develop.2020.7.29+commit.f2fa5b5f.Linux.clang
- Target EVM version (as per compiler settings): N/A
- Framework/IDE (e.g. Truffle or Remix): N/A
- EVM execution environment / backend / blockchain client: N/A
- Operating system: Ubuntu 18.04 in docker

## Steps to Reproduce

Above shows pretty clearly, I think.

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/ethereum/solidity/issues/9196
Fixes https://github.com/ethereum/solidity/issues/9550

### CODE REVIEW

File: test/libsolidity/syntaxTests/immutable/unary.sol [hunk 1]
```diff
@@ -0,0 +1,9 @@
+contract A {
+  int immutable a;
+  constructor() { a = 5; }
+  function f() public { --a; }
+}
+
+// ----
+// TypeError 1581: (85-86): Immutable variables can only be initialized inline or assigned directly in the constructor.
+// TypeError 1574: (85-86): Immutable state variable already initialized.
```
Comment (leonardoalt) [inline test/libsolidity/syntaxTests/immutable/unary.sol]: This is error is not really bad, but kinda misleading.

Comment (chriseth) [reply test/libsolidity/syntaxTests/immutable/unary.sol]: I will check if I can improve it. NB: it should complain about the writing, not about the re-writing.

**Comment by chriseth** [comment]
Rebased and updated. I also changed the wording to be more explicit that this is about writing.

**Comment by chriseth** [comment]
Tests are green, but not reported somehow.

**Comment by axic** [comment]
> Tests are green, but not reported somehow.

They seem to be now.

**Review by leonardoalt — COMMENTED** [review]
Looks good.

Maybe add a test like
```
uint immutable x;
uint immutable y;
constructor() {
  ++x;
  --y;
}
```

**Comment by chriseth** [comment]
added test.

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**